rasterio/io.py

```diff
@@ -170,6 +170,32 @@ def __repr__(self):
         return "<{} DatasetReader name='{}' mode='{}'>".format(
             self.closed and 'closed' or 'open', self.name, self.mode)
 
+    @property
+    def mask_flag_enums(self):
+        """Sets of flags describing the sources of band masks.
+
+        Returns
+        -------
+        list [, list*]
+            One list of rasterio.enums.MaskFlags members per band.
+
+        Examples
+        --------
+
+        For a 3 band dataset that has masks derived from nodata values:
+
+        >>> dataset.mask_flag_enums
+        ([<MaskFlags.nodata: 8>], [<MaskFlags.nodata: 8>], [<MaskFlags.nodata: 8>])
+        >>> band1_flags = dataset.mask_flag_enums[0]
+        >>> rio.enums.MaskFlags.nodata in band1_flags
+        True
+        >>> rio.enums.MaskFlags.alpha in band1_flags
+        False
+
+        """
+        return tuple(
+            [flag for flag in enums.MaskFlags if x & flag.value]
+            for x in self.mask_flags)
```

Comment:
Maybe we should document here the possible flags and what they mean. The terminology is a bit confusing unless you're familiar with GDAL rfc15. We could just crib the descriptions from there:

```
all_valid: There are no invalid pixels, all mask values will be 255. When used this will normally be the only flag set.
per_dataset: The mask band is shared between all bands on the dataset.
alpha: The mask band is actually an alpha band and may have values other than 0 and 255.
nodata: Indicates the mask is actually being generated from nodata values. (mutually exclusive of "alpha")
```